piet-common/src/cairo_back.rs

```diff
@@ -101,33 +101,60 @@ impl<'a> BitmapTarget<'a> {
         CairoRenderContext::new(&mut self.cr)
     }
 
-    /// Get raw RGBA pixels from the bitmap.
-    pub fn into_raw_pixels(mut self, fmt: ImageFormat) -> Result<Vec<u8>, piet::Error> {
+    /// Get raw RGBA pixels from the bitmap by copying them into `buf`. If all the pixels were
+    /// copied, returns the number of bytes written. If `buf` wasn't big enough, returns an error
+    /// and doesn't write anything.
+    pub fn copy_raw_pixels(
+        &mut self,
+        fmt: ImageFormat,
+        buf: &mut [u8],
+    ) -> Result<usize, piet::Error> {
         // TODO: convert other formats.
         if fmt != ImageFormat::RgbaPremul {
             return Err(piet::Error::NotSupported);
         }
-        std::mem::drop(self.cr);
         self.surface.flush();
         let stride = self.surface.get_stride() as usize;
         let width = self.surface.get_width() as usize;
         let height = self.surface.get_height() as usize;
-        let mut raw_data = vec![0; width * height * 4];
-        let buf = self
-            .surface
-            .get_data()
-            .map_err(Into::<Box<dyn std::error::Error>>::into)?;
-        for y in 0..height {
-            let src_off = y * stride;
-            let dst_off = y * width * 4;
-            for x in 0..width {
-                raw_data[dst_off + x * 4 + 0] = buf[src_off + x * 4 + 2];
-                raw_data[dst_off + x * 4 + 1] = buf[src_off + x * 4 + 1];
-                raw_data[dst_off + x * 4 + 2] = buf[src_off + x * 4 + 0];
-                raw_data[dst_off + x * 4 + 3] = buf[src_off + x * 4 + 3];
-            }
+        let size = width * height * 4;
+        if buf.len() < size {
+            return Err(piet::Error::InvalidInput);
         }
-        Ok(raw_data)
+        unsafe {
+            // Cairo's rust wrapper has extra safety checks that we want to avoid: it won't let us
```

Comment:
I would open an issue upstream and see if cairo-rs is open to adding API for this.

For what it's worth, I believe it was the lack of this API that is the reason this functionality was not originally implemented.

Reply:
I'll go do that now.

Reply:
Also: I *am* okay with having the unsafe here now, but I would include a link to the docs that support the assertion that immutable access is okay.